### PR TITLE
Improve keyboard-teleop and datalogger

### PR DESCRIPTION
**Pull request by KavitShah1998**

### CODE REVIEW

File: bd_spot_wrapper/spot_wrapper/data_logger.py [hunk 1]
```diff
@@ -104,41 +96,220 @@ def convert_depth_to_img(raw_depth):
...
+
+    def setup_logging_sources(self, camera_sources: List[str]):
+        """
+        Order .. RGB & then DEPTH_IN_RGB
+        DO NOT SUPPORT DEPTH and RGB_IN_DEPTH.
+        """
+        # By default, always log for hand camera data
+        source_list = [
+            SpotCamIds.INTEL_REALSENSE_COLOR,
+            SpotCamIds.INTEL_REALSENSE_DEPTH,
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/data_logger.py:139]: Update this to use hand-camera by default

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py [hunk 2]
```diff
@@ -23,18 +24,30 @@
...
 INITIAL_POINT = np.array([0.5, 0.0, 0.35])
-INITIAL_RPY = np.deg2rad([0.0, 45.0, 0.0])
+INITIAL_RPY = np.deg2rad([0.0, np.pi / 4, 0.0])
+
+# TODO: Use the joint states from config.yaml
+# INITIAL_POINT_GRIPPERCAM = np.array([0.5, 0.0, 0.81])
+# INITIAL_RPY_GRIPPERCAM = np.deg2rad([0.0, np.pi/4, 0.0])
+
+# INITIAL_POINT_INTELCAM = np.array([0.5, 0.0, 0.85])
+# INITIAL_RPY_INTELCAM = np.deg2rad([0.0, np.pi/3, 0.0])
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:34]: Remove these changes

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py [hunk 3]
```diff
@@ -74,23 +102,48 @@ def raise_error(sig, frame):
...
+            y_pos=y0,
+            yaw=theta0 + (i + 1) * 2 * np.pi / n_intervals,
+            end_time=100,
+            blocking=True,
+        )
+        datalogger.log_data_finite(n_captures)
+
+
+# @click.command()
+# @click.option("--initial-arm-state", type=int, help="Use 0 for default, 1 for gripperCam logger, and 2 for intelRealsense logger")
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:119]: Remove this

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py [hunk 4]
```diff
@@ -74,23 +102,48 @@ def raise_error(sig, frame):
...
+    # TODO: Add code for moving arm to higher perspective for better view
     # Move arm to initial configuration
-    point, rpy = move_to_initial(spot)
+    point, rpy = move_to_initial(spot, initial_arm_state)
     control_arm = False
 
+    enable_logger_during_teleop = False
+
+    # Rotate robot after undocking and capture data
+    # rotate(datalogger=datalogger)
```
Comment (KavitShah1998) [inline bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:145]: Uncomment this

File: intel_realsense_payload_for_spotsim2real/caliberate_intel_with_gripper_cam_icp.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    # Apply ICP
+    icp_result = o3d.pipelines.registration.registration_icp(
+        intel_pcd_full,
+        gripper_pcd_full,
+        max_correspondence_distance=0.05,
+        init=transformation,
+        estimation_method=o3d.pipelines.registration.TransformationEstimationPointToPoint(),
+    )
+
+    print("Refined Transformation Matrix:\n", icp_result.transformation)
```
Comment (KavitShah1998) [inline intel_realsense_payload_for_spotsim2real/caliberate_intel_with_gripper_cam_icp.py:204]: Remove this file

File: perception_and_utils_root/perception_and_utils/perception/detector_wrappers/april_tag_detector.py [hunk 6]
```diff
@@ -156,6 +156,7 @@ def get_outputs(
                 - tag_image_metadata_list (List[Any]) : List of image metadata
 
         """
+        print(type(img_frame))
```
Comment (KavitShah1998) [inline perception_and_utils_root/perception_and_utils/perception/detector_wrappers/april_tag_detector.py:159]: Remove this

File: spot_rl_experiments/configs/data_waypoints.json [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      "y": 3.6068365395073307
+    }
+  },
+  {
+    "20": {
+      "x": 5.007439059032365,
+      "y": 3.702357202263098
+    }
+  }
+]
```
Comment (KavitShah1998) [inline spot_rl_experiments/configs/data_waypoints.json:110]: Remove this,

File: spot_rl_experiments/configs/data_waypoints_copied.json [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      "y": 7.5723976450621215
+    }
+  },
+  {
+    "29": {
+      "x": 5.4056724764130895,
+      "y": 7.963035689017092
+    }
+  }
+]
```
Comment (KavitShah1998) [inline spot_rl_experiments/configs/data_waypoints_copied.json:176]: Remove this

File: spot_rl_experiments/spot_rl/utils/search_table_location.py [hunk 9]
```diff
@@ -9,13 +9,14 @@
...
 from spot_rl.envs.base_env import SpotBaseEnv
+from spot_rl.utils.gripper_T_intel import GRIPPER_T_INTEL
 from spot_rl.utils.heuristic_nav import get_3d_point, get_best_uvz_from_detection
 from spot_rl.utils.plane_detection import plane_detect
 from spot_rl.utils.utils import ros_topics as rt
 from spot_wrapper.spot import Spot, image_response_to_cv2
 from std_msgs.msg import String
 
-GRIPPER_T_INTEL = osp.join(osp.dirname(osp.abspath(__file__)), "gripper_T_intel.npy")
+# GRIPPER_T_INTEL = osp.join(osp.dirname(osp.abspath(__file__)), "gripper_T_intel.npy")
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/utils/search_table_location.py:19]: Remove this

File: spot_rl_experiments/spot_rl/utils/utils.py [hunk 10]
```diff
@@ -167,3 +168,30 @@ def __setitem__(self, key, value):
...
+            x, y, _ = values
+            all_waypoints.append({key: {"x": x, "y": y}})
+
+    # Write the combined JSON data to a file
+    with open(DATA_WAYPOINTS_JSON, 'w') as file:
+        json.dump(all_waypoints, file, indent=2)
+
+    print("Combined JSON data has been written to 'all_waypoints.json'.")
+
+if __name__ == "__main__":
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/utils/utils.py:196]: This should also be a part of separate PR along with `create_PCD_from_waypoints.py`

File: bd_spot_wrapper/spot_wrapper/data_logger.py (hunk 1 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/data_logger.py:139]: done

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py (hunk 2 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:34]: done

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py (hunk 4 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:145]: done

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py (hunk 3 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:119]: done

File: perception_and_utils_root/perception_and_utils/perception/detector_wrappers/april_tag_detector.py (hunk 6 above)
Comment (KavitShah1998) [reply perception_and_utils_root/perception_and_utils/perception/detector_wrappers/april_tag_detector.py:159]: done

File: spot_rl_experiments/spot_rl/utils/search_table_location.py (hunk 9 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/utils/search_table_location.py:19]: done

File: intel_realsense_payload_for_spotsim2real/caliberate_intel_with_gripper_cam_icp.py (hunk 5 above)
Comment (KavitShah1998) [reply intel_realsense_payload_for_spotsim2real/caliberate_intel_with_gripper_cam_icp.py]: removed

File: spot_rl_experiments/configs/data_waypoints_copied.json (hunk 8 above)
Comment (KavitShah1998) [reply spot_rl_experiments/configs/data_waypoints_copied.json]: removed

File: spot_rl_experiments/spot_rl/utils/utils.py (hunk 10 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/utils/utils.py]: removed

File: spot_rl_experiments/configs/data_waypoints.json (hunk 7 above)
Comment (KavitShah1998) [reply spot_rl_experiments/configs/data_waypoints.json]: removed

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py [hunk 11]
```diff
@@ -23,18 +24,24 @@
 
 # Where the gripper goes to upon initialization
 INITIAL_POINT = np.array([0.5, 0.0, 0.35])
-INITIAL_RPY = np.deg2rad([0.0, 45.0, 0.0])
+INITIAL_RPY = np.deg2rad([0.0, np.pi / 4, 0.0])
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:27]: HI Kavit, it seems that you want to have pi/4 rad = 45 deg, But there is a np.deg2rad() outside the list

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 12]
```diff
@@ -1239,7 +1150,7 @@ def dock(self, dock_id: int = DOCK_ID, home_robot: bool = False) -> None:
     def undock(self):
         # Before undocking from dock, update the global_T_home transform
         # member variable but not don't persist the changes in home.txt
-        self.home_robot(write_to_file=False)
+        # self.home_robot(write_to_file=False)
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/spot.py:1153]: no update here?

**Review by jimmytyyang — COMMENTED** [review]
good PR, kavit! I left some easy-to-fix comments. I pushed a small commit to your PR.

**Comment by jimmytyyang** [comment]
@KavitShah1998 Please git pull when you address the comments

File: bd_spot_wrapper/spot_wrapper/spot.py (hunk 12 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/spot.py:1153]: This was actually source of a bug. Thanks for pointing this out. WIll remove it.

File: bd_spot_wrapper/spot_wrapper/keyboard_teleop.py (hunk 11 above)
Comment (KavitShah1998) [reply bd_spot_wrapper/spot_wrapper/keyboard_teleop.py:27]: Makes sense, good find.
Addressed.

**Review by jimmytyyang — APPROVED** [review]
LGTM!
